packages/run-protocol/src/vpool-xyk-amm/addPool.js

```diff
@@ -12,50 +13,92 @@ const { details: X } = assert;
  * @param {ZCF} zcf
  * @param {(Brand) => boolean} isInSecondaries
  * @param {WeakStore<Brand,ZCFMint>} brandToLiquidityMint
+ * @param {WeakStore<Brand,ERef<Issuer>>} brandToLiquidityIssuer
  * @param {() => (secondaryBrand: Brand) => Promise<void>} getAddIssuerToReserve
  */
 export const makeAddIssuer = (
   zcf,
   isInSecondaries,
   brandToLiquidityMint,
+  brandToLiquidityIssuer,
   getAddIssuerToReserve,
 ) => {
   /**
+   * Add a new issuer. If we previously received a request for the same issuer,
+   * we'll return the issuer or a promise for it.
+   *
    * @param {Issuer} secondaryIssuer
    * @param {string} keyword
    */
   return async (secondaryIssuer, keyword) => {
-    const [secondaryAssetKind, secondaryBrand] = await Promise.all([
-      E(secondaryIssuer).getAssetKind(),
-      E(secondaryIssuer).getBrand(),
-    ]);
-
-    assert(
-      !isInSecondaries(secondaryBrand),
-      X`issuer ${secondaryIssuer} already has a pool`,
-    );
-    assert(
-      secondaryAssetKind === AssetKind.NAT,
-      X`${keyword} asset not fungible (must use NAT math)`,
-    );
-    const liquidityKeyword = `${keyword}Liquidity`;
-    zcf.assertUniqueKeyword(liquidityKeyword);
-
-    const mint = await zcf.makeZCFMint(
-      liquidityKeyword,
-      AssetKind.NAT,
-      harden({ decimalPlaces: 6 }),
+    return E.when(
+      Promise.all([
+        E(secondaryIssuer).getAssetKind(),
+        E(secondaryIssuer).getBrand(),
+      ]),
+      ([secondaryAssetKind, secondaryBrand]) => {
+        assert(
+          secondaryAssetKind === AssetKind.NAT,
+          X`${keyword} asset not fungible (must use NAT math)`,
+        );
+
+        if (brandToLiquidityIssuer.has(secondaryBrand)) {
+          return brandToLiquidityIssuer.get(secondaryBrand);
+        }
+
+        assert(
+          !isInSecondaries(secondaryBrand),
+          X`issuer ${secondaryIssuer} already has a pool`,
+        );
+
+        if (brandToLiquidityMint.has(secondaryBrand)) {
+          const { issuer } = brandToLiquidityMint
+            .get(secondaryBrand)
+            .getIssuerRecord();
+          return issuer;
+        }
+
+        /** @template T @typedef {import('@endo/promise-kit').PromiseKit<T>} PromiseKit */
+        /** @type {PromiseKit<Issuer<'nat'>>} */
+        const promiseKit = makePromiseKit();
+        brandToLiquidityIssuer.init(secondaryBrand, promiseKit.promise);
```

Comment:
At this point, we've taken the lock. I believe that this doesn't get correctly pulled in the error case.